packages/plugins/response-cache/src/plugin.ts

```diff
@@ -205,63 +206,77 @@ export type ResponseCacheExecutionResult = ExecutionResult<
   { responseCache?: ResponseCacheExtensions }
 >;
 
-const originalDocumentMap = new WeakMap<DocumentNode, DocumentNode>();
-const addEntityInfosToDocument = memoize4(function addTypeNameToDocument(
+const getDocumentWithMetadataAndTTL = memoize4(function addTypeNameToDocument(
   document: DocumentNode,
-  addTypeNameToDocumentOpts: { invalidateViaMutation: boolean },
+  {
+    invalidateViaMutation,
+    ttlPerSchemaCoordinate,
+  }: {
+    invalidateViaMutation: boolean;
+    ttlPerSchemaCoordinate?: Record<string, number | undefined>;
+  },
   schema: any,
   idFieldByTypeName: Map<string, string>,
-): DocumentNode {
+): [DocumentNode, number | undefined] {
   const typeInfo = new TypeInfo(schema);
-  const newDocument = visit(
-    document,
-    visitWithTypeInfo(typeInfo, {
-      OperationDefinition: {
-        enter(node): void | false {
-          if (!addTypeNameToDocumentOpts.invalidateViaMutation && node.operation === 'mutation') {
-            return false;
-          }
-          if (node.operation === 'subscription') {
-            return false;
-          }
-        },
+  let ttl: number | undefined;
+  const visitor: ASTVisitor = {
+    OperationDefinition: {
+      enter(node): void | false {
+        if (!invalidateViaMutation && node.operation === 'mutation') {
+          return false;
+        }
+        if (node.operation === 'subscription') {
+          return false;
+        }
       },
-      SelectionSet(node, _key) {
+    },
+    ...(ttlPerSchemaCoordinate != null && {
+      Field(fieldNode) {
         const parentType = typeInfo.getParentType();
-        const idField = parentType && idFieldByTypeName.get(parentType.name);
-        return {
-          ...node,
-          selections: [
-            {
-              kind: Kind.FIELD,
-              name: {
-                kind: Kind.NAME,
-                value: '__typename',
-              },
-              alias: {
-                kind: Kind.NAME,
-                value: '__responseCacheTypeName',
-              },
-            },
-            ...(idField
-              ? [
-                  {
-                    kind: Kind.FIELD,
-                    name: { kind: Kind.NAME, value: idField },
-                    alias: { kind: Kind.NAME, value: '__responseCacheId' },
-                  },
-                ]
-              : []),
-
-            ...node.selections,
-          ],
-        };
+        if (parentType) {
```

Comment:
Do you think we can skip ttl calculation for mutations?

Reply:
That's what I do :-) But it's not very obvious... When it's a mutation, I don't give the `ttlPerSchemaCoordinate` parameter, so this visitor is not added.

Reply:
Ah ok! Perfect then. I thought we never had this before :D